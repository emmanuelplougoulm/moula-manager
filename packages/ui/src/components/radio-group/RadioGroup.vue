<template>
  <InputGroupContainer
    :label="label"
    :tooltip-title="tooltipTitle"
    :tooltip-description="tooltipDescription"
    :selected="!!selectedOption"
  >
    <div v-for="option in options" :key="option.id">
      <input
        :id="option.value"
        v-model="selectedOption"
        type="radio"
        :value="option.value"
        :name="name"
        :data-testid="`input-${option.id}`"
        :disabled="option.disabled"
        @change="onChange"
      />
      <label class="text-big-regular" :for="option.value">
        <div class="icon-container">
          <RadioButtonChecked v-if="checked(option)" class="icon" />
          <RadioButtonUnchecked v-else class="icon" />
        </div>
        {{ option.text }}
      </label>
    </div>
  </InputGroupContainer>
</template>

<script setup lang="ts">
import { ref } from 'vue';
import InputGroupContainer from '../input-group-container/InputGroupContainer.vue';
import RadioButtonChecked from '../icons/RadioButtonChecked.vue';
import RadioButtonUnchecked from '../icons/RadioButtonUnchecked.vue';
import type { InputOptionType } from '../../types';

const props = defineProps({
  modelValue: {
    type: String,
    required: true
  },
  options: {
    type: Array as () => InputOptionType[],
    required: true
  },
  name: {
    type: String,
    required: true
  },
  label: {
    type: String,
    required: true
  },
  tooltipTitle: {
    type: String,
    default: undefined
  },
  tooltipDescription: {
    type: String,
    default: undefined
  }
});

const selectedOption = ref(props.modelValue);
const checked = (option: InputOptionType) => selectedOption.value === option.value;

const emit = defineEmits(['update:modelValue']);

const onChange = () => {
  emit('update:modelValue', selectedOption.value);
};
</script>

<style scoped>
input {
  display: none;
}

label {
  display: flex;
  align-items: center;
  color: var(--neutral-20);
  background-color: var(--neutral-100);
  cursor: pointer;
  padding: 0.8125rem 1rem;
  gap: 0.5rem;
  line-height: 1.375rem;
  outline: 0.063rem solid var(--neutral-60);
  border-radius: 0.25rem;
}

label:hover {
  background-color: var(--neutral-80);
}

input:checked + label {
  outline: 0.125rem solid var(--neutral-20);
  border-radius: 0.125rem;
  background-color: var(--neutral-80);
}

input:disabled + label {
  opacity: 0.3;
}

.icon-container {
  display: flex;
  align-items: center;
  justify-content: center;
  transition: all 0.2s;
  box-sizing: border-box;
}

.icon {
  color: var(--neutral-40);
  width: 1.375rem;
  height: 1.375rem;
}
</style>
