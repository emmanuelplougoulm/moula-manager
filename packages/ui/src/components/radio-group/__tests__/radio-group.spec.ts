import { mount } from '@vue/test-utils';
import Icons from '../../icons/index';
import RadioGroup from '../RadioGroup.vue';

const inputTestId = "[data-testid='input-1']";

const options = [
  {
    id: 0,
    text: 'Option 1',
    value: 'option-1',
    disabled: false
  },
  {
    id: 1,
    text: 'Option 2',
    value: 'option-2',
    disabled: false
  }
];

const disabledOptions = [
  {
    id: 0,
    text: 'Option 1',
    value: 'option-1',
    disabled: true
  },
  {
    id: 1,
    text: 'Option 2',
    value: 'option-2',
    disabled: true
  }
];

describe('RadioGroup tests', () => {
  it('should display a highlighted label and no error message by default', async () => {
    const wrapper = mount(RadioGroup, {
      global: {
        plugins: [Icons]
      },
      props: {
        label: 'Text',
        name: 'name',
        options: options,
        modelValue: ''
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display a non highlighted label and no error message if an option is checked', async () => {
    const wrapper = mount(RadioGroup, {
      global: {
        plugins: [Icons]
      },
      props: {
        label: 'Text',
        name: 'name',
        error: 'This field is required',
        options: options,
        modelValue: ''
      }
    });
    const input = wrapper.find(inputTestId);
    await input.setValue(true);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display the component with disabled options', () => {
    const wrapper = mount(RadioGroup, {
      global: {
        plugins: [Icons]
      },
      props: {
        label: 'Text',
        options: disabledOptions,
        name: 'name',
        modelValue: ''
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should emit an event when a box is checked', async () => {
    const wrapper = mount(RadioGroup, {
      props: {
        label: 'Text',
        options: options,
        name: 'name',
        modelValue: ''
      }
    });
    const input = wrapper.find(inputTestId);
    await input.setValue(true);
    expect(wrapper.emitted()).toHaveProperty('update:modelValue');
  });
});
