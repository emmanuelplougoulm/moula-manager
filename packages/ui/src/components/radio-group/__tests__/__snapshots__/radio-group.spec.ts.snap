// Vitest Snapshot v1

exports[`RadioGroup tests > should display a highlighted label and no error message by default 1`] = `
"<div data-v-35f5b373=\\"\\" data-v-e142b43e=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-35f5b373=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Text <!--v-if--></span>
  <div data-v-e142b43e=\\"\\"><input data-v-e142b43e=\\"\\" id=\\"option-1\\" type=\\"radio\\" name=\\"name\\" data-testid=\\"input-0\\" value=\\"option-1\\"><label data-v-e142b43e=\\"\\" class=\\"text-big-regular\\" for=\\"option-1\\">
      <div data-v-e142b43e=\\"\\" class=\\"icon-container\\">
        <div data-v-b9dd0015=\\"\\" data-v-e142b43e=\\"\\" class=\\"container icon\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M480.078 990.131q-85.469 0-161.006-32.395-75.536-32.395-131.975-88.833-56.438-56.439-88.833-131.897-32.395-75.459-32.395-160.928 0-86.469 32.395-162.006 32.395-75.536 88.745-131.504 56.349-55.968 131.849-88.616 75.5-32.648 161.017-32.648 86.516 0 162.12 32.604 75.603 32.604 131.529 88.497t88.549 131.452Q894.696 489.416 894.696 576q0 85.547-32.648 161.075-32.648 75.527-88.616 131.896-55.968 56.37-131.426 88.765-75.459 32.395-161.928 32.395ZM480 910.913q139.739 0 237.326-97.732Q814.913 715.449 814.913 576q0-139.739-97.587-237.326-97.587-97.587-237.609-97.587-139.021 0-236.826 97.587-97.804 97.587-97.804 237.609 0 139.021 97.732 236.826Q340.551 910.913 480 910.913ZM480 576Z\\"></path>
          </svg></div>
      </div> Option 1
    </label></div>
  <div data-v-e142b43e=\\"\\"><input data-v-e142b43e=\\"\\" id=\\"option-2\\" type=\\"radio\\" name=\\"name\\" data-testid=\\"input-1\\" value=\\"option-2\\"><label data-v-e142b43e=\\"\\" class=\\"text-big-regular\\" for=\\"option-2\\">
      <div data-v-e142b43e=\\"\\" class=\\"icon-container\\">
        <div data-v-b9dd0015=\\"\\" data-v-e142b43e=\\"\\" class=\\"container icon\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M480.078 990.131q-85.469 0-161.006-32.395-75.536-32.395-131.975-88.833-56.438-56.439-88.833-131.897-32.395-75.459-32.395-160.928 0-86.469 32.395-162.006 32.395-75.536 88.745-131.504 56.349-55.968 131.849-88.616 75.5-32.648 161.017-32.648 86.516 0 162.12 32.604 75.603 32.604 131.529 88.497t88.549 131.452Q894.696 489.416 894.696 576q0 85.547-32.648 161.075-32.648 75.527-88.616 131.896-55.968 56.37-131.426 88.765-75.459 32.395-161.928 32.395ZM480 910.913q139.739 0 237.326-97.732Q814.913 715.449 814.913 576q0-139.739-97.587-237.326-97.587-97.587-237.609-97.587-139.021 0-236.826 97.587-97.804 97.587-97.804 237.609 0 139.021 97.732 236.826Q340.551 910.913 480 910.913ZM480 576Z\\"></path>
          </svg></div>
      </div> Option 2
    </label></div>
  <!--v-if-->
</div>"
`;

exports[`RadioGroup tests > should display a non highlighted label and no error message if an option is checked 1`] = `
"<div data-v-35f5b373=\\"\\" data-v-e142b43e=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-35f5b373=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default\\" data-testid=\\"label-item\\">Text <!--v-if--></span>
  <div data-v-e142b43e=\\"\\"><input data-v-e142b43e=\\"\\" id=\\"option-1\\" type=\\"radio\\" name=\\"name\\" data-testid=\\"input-0\\" value=\\"option-1\\"><label data-v-e142b43e=\\"\\" class=\\"text-big-regular\\" for=\\"option-1\\">
      <div data-v-e142b43e=\\"\\" class=\\"icon-container\\">
        <div data-v-b9dd0015=\\"\\" data-v-e142b43e=\\"\\" class=\\"container icon\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M480.078 990.131q-85.469 0-161.006-32.395-75.536-32.395-131.975-88.833-56.438-56.439-88.833-131.897-32.395-75.459-32.395-160.928 0-86.469 32.395-162.006 32.395-75.536 88.745-131.504 56.349-55.968 131.849-88.616 75.5-32.648 161.017-32.648 86.516 0 162.12 32.604 75.603 32.604 131.529 88.497t88.549 131.452Q894.696 489.416 894.696 576q0 85.547-32.648 161.075-32.648 75.527-88.616 131.896-55.968 56.37-131.426 88.765-75.459 32.395-161.928 32.395ZM480 910.913q139.739 0 237.326-97.732Q814.913 715.449 814.913 576q0-139.739-97.587-237.326-97.587-97.587-237.609-97.587-139.021 0-236.826 97.587-97.804 97.587-97.804 237.609 0 139.021 97.732 236.826Q340.551 910.913 480 910.913ZM480 576Z\\"></path>
          </svg></div>
      </div> Option 1
    </label></div>
  <div data-v-e142b43e=\\"\\"><input data-v-e142b43e=\\"\\" id=\\"option-2\\" type=\\"radio\\" name=\\"name\\" data-testid=\\"input-1\\" value=\\"option-2\\"><label data-v-e142b43e=\\"\\" class=\\"text-big-regular\\" for=\\"option-2\\">
      <div data-v-e142b43e=\\"\\" class=\\"icon-container\\">
        <div data-v-b9dd0015=\\"\\" data-v-e142b43e=\\"\\" class=\\"container icon\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M480 762q78 0 132-54t54-132q0-78-54-132t-132-54q-78 0-132 54t-54 132q0 78 54 132t132 54Zm.078 228.131q-85.469 0-161.006-32.395-75.536-32.395-131.975-88.833-56.438-56.439-88.833-131.897-32.395-75.459-32.395-160.928 0-86.469 32.395-162.006 32.395-75.536 88.745-131.504 56.349-55.968 131.849-88.616 75.5-32.648 161.017-32.648 86.516 0 162.12 32.604 75.603 32.604 131.529 88.497t88.549 131.452Q894.696 489.416 894.696 576q0 85.547-32.648 161.075-32.648 75.527-88.616 131.896-55.968 56.37-131.426 88.765-75.459 32.395-161.928 32.395ZM480 910.913q139.739 0 237.326-97.732Q814.913 715.449 814.913 576q0-139.739-97.587-237.326-97.587-97.587-237.609-97.587-139.021 0-236.826 97.587-97.804 97.587-97.804 237.609 0 139.021 97.732 236.826Q340.551 910.913 480 910.913ZM480 576Z\\"></path>
          </svg></div>
      </div> Option 2
    </label></div><span data-v-c1d8433f=\\"\\" data-v-35f5b373=\\"\\" id=\\"error-item\\" data-testid=\\"error-item\\" class=\\"text-small-regular\\">This field is required</span>
</div>"
`;

exports[`RadioGroup tests > should display the component with disabled options 1`] = `
"<div data-v-35f5b373=\\"\\" data-v-e142b43e=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-35f5b373=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Text <!--v-if--></span>
  <div data-v-e142b43e=\\"\\"><input data-v-e142b43e=\\"\\" id=\\"option-1\\" type=\\"radio\\" name=\\"name\\" data-testid=\\"input-0\\" disabled=\\"\\" value=\\"option-1\\"><label data-v-e142b43e=\\"\\" class=\\"text-big-regular\\" for=\\"option-1\\">
      <div data-v-e142b43e=\\"\\" class=\\"icon-container\\">
        <div data-v-b9dd0015=\\"\\" data-v-e142b43e=\\"\\" class=\\"container icon\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M480.078 990.131q-85.469 0-161.006-32.395-75.536-32.395-131.975-88.833-56.438-56.439-88.833-131.897-32.395-75.459-32.395-160.928 0-86.469 32.395-162.006 32.395-75.536 88.745-131.504 56.349-55.968 131.849-88.616 75.5-32.648 161.017-32.648 86.516 0 162.12 32.604 75.603 32.604 131.529 88.497t88.549 131.452Q894.696 489.416 894.696 576q0 85.547-32.648 161.075-32.648 75.527-88.616 131.896-55.968 56.37-131.426 88.765-75.459 32.395-161.928 32.395ZM480 910.913q139.739 0 237.326-97.732Q814.913 715.449 814.913 576q0-139.739-97.587-237.326-97.587-97.587-237.609-97.587-139.021 0-236.826 97.587-97.804 97.587-97.804 237.609 0 139.021 97.732 236.826Q340.551 910.913 480 910.913ZM480 576Z\\"></path>
          </svg></div>
      </div> Option 1
    </label></div>
  <div data-v-e142b43e=\\"\\"><input data-v-e142b43e=\\"\\" id=\\"option-2\\" type=\\"radio\\" name=\\"name\\" data-testid=\\"input-1\\" disabled=\\"\\" value=\\"option-2\\"><label data-v-e142b43e=\\"\\" class=\\"text-big-regular\\" for=\\"option-2\\">
      <div data-v-e142b43e=\\"\\" class=\\"icon-container\\">
        <div data-v-b9dd0015=\\"\\" data-v-e142b43e=\\"\\" class=\\"container icon\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M480.078 990.131q-85.469 0-161.006-32.395-75.536-32.395-131.975-88.833-56.438-56.439-88.833-131.897-32.395-75.459-32.395-160.928 0-86.469 32.395-162.006 32.395-75.536 88.745-131.504 56.349-55.968 131.849-88.616 75.5-32.648 161.017-32.648 86.516 0 162.12 32.604 75.603 32.604 131.529 88.497t88.549 131.452Q894.696 489.416 894.696 576q0 85.547-32.648 161.075-32.648 75.527-88.616 131.896-55.968 56.37-131.426 88.765-75.459 32.395-161.928 32.395ZM480 910.913q139.739 0 237.326-97.732Q814.913 715.449 814.913 576q0-139.739-97.587-237.326-97.587-97.587-237.609-97.587-139.021 0-236.826 97.587-97.804 97.587-97.804 237.609 0 139.021 97.732 236.826Q340.551 910.913 480 910.913ZM480 576Z\\"></path>
          </svg></div>
      </div> Option 2
    </label></div>
  <!--v-if-->
</div>"
`;
